chainer/gradient_check.py

```diff
@@ -52,34 +85,182 @@ def numerical_grad(f, inputs, grad_outputs, eps=1e-3):
         )
     else:
         xp = numpy
     grads = [xp.zeros(x.shape, numpy.float64) for x in inputs]
 
+    if detect_nondifferentiable:
+        if center_outputs is None:
+            ys0 = _copy_arrays(f())
+        else:
+            ys0 = center_outputs
+        nout = len(ys0)
+        shapes = [_.shape for _ in ys0]
+        sizes = numpy.array([_.size for _ in ys0])
+        cumsizes = numpy.cumsum(sizes)
+
+    # Evaluate func at a single input
+    def eval_func(x, i, delta, orig):
+        x[i] = orig + delta
+        y = _copy_arrays(f())
+        x[i] = orig
+        return y
+
+    # An iteration on a single input displacement
+    def iterate_single_input(i_in, x, orig_x, i):
+        orig = orig_x[i]
+        if detect_nondifferentiable:
+            yss = [
+                eval_func(x, i, -eps * 1., orig),
+                eval_func(x, i, -eps * .5, orig),
+                ys0,
+                eval_func(x, i, +eps * .5, orig),
+                eval_func(x, i, +eps * 1., orig),
+            ]
+        else:
+            yss = [
+                eval_func(x, i, -eps * 1, orig),
+                eval_func(x, i, +eps * 1, orig),
+            ]
+
+        if detect_nondifferentiable:
+            # Detect non-differentiable point by quadratic fitting
+
+            # Check for non-finite output.
+            # If any single element in the output arrays has different
+            # finiteness among sampled points, that means this is a
+            # non-differentiable point.
+            # If the function consistently generates non-finite values
+            # around the point, we do not treat the point as
+            # non-differentiable.
+            # (Example: x<0 region for the logarithm function)
+            any_nonfinite = False
+            for i_out in range(nout):
+                isfinites = [
+                    xp.isfinite(yss[_][i_out])
+                    for _ in range(len(yss))]
+                if any((isfinites[0] != isfinites[_]).any()
+                       for _ in range(1, len(yss))):
+                    s = six.StringIO()
+                    s.write(
+                        'Tried to compute the numeric gradient on a '
+                        'non-differentiable point.\n\n')
+                    s.write('i_in: {}\n'.format(i_in))
+                    s.write('i_out: {}\n'.format(i_out))
+                    s.write('x: {}\n'.format(inputs[i_in]))
+                    s.write('index on x: {}\n'.format(i))
+                    s.write('eps: {}\n'.format(eps))
+                    s.write('y[x-eps  ]: {}\n'.format(yss[0][i_out]))
+                    s.write('y[x-eps/2]: {}\n'.format(yss[1][i_out]))
+                    s.write('y[x      ]: {}\n'.format(yss[2][i_out]))
+                    s.write('y[x+eps/2]: {}\n'.format(yss[3][i_out]))
+                    s.write('y[x+eps  ]: {}\n'.format(yss[4][i_out]))
+                    raise NondifferentiableError(s.getvalue())
```

Comment:
Do we need to close the file, or is it being garbage collected after returning from the function?

Reply:
`s` is `StringIO`, not file. It will be garbage collected.

Reply:
No change required.